### ISSUE TITLE
SearchableAttributes change order of fields in response hits

### ISSUE THREAD

**Issue by bidoubiwa** [meilisearch/MeiliSearch#1495]
**Describe the bug**

When adding fields in `searchableAttributes` the order of fields in the returned document are impacted:
```json
{
  "searchableAttributes": [
    "title",
    "description"
  ],
  "filterableAttributes": [
    "genres"
  ]
}
```

`id` field of document goes from first position to second position.
Given this document: 

```json
{
  "id": 28,
  "title": "Apocalypse Now",
  "overview": "At the height of the Vietnam war, Captain Benjamin Willard is sent on a dangerous mission that, officially, \"does not exist, nor will it ever exist.\" His goal is to locate - and eliminate - a mysterious Green Beret Colonel named Walter Kurtz, who has been leading his personal army on illegal guerrilla missions into enemy territory.",
  "genres": [
    "Drama",
    "War"
  ],
  "poster": "https://image.tmdb.org/t/p/w500/gQB8Y5RCMkv2zwzFHbUJX3kAhvA.jpg",
  "release_date": 303523200
}
```

after a search this is how it is returned in the response body: 

```json
        {
            "title": "Apocalypse Now",
            "id": 28,
            "overview": "At the height of the Vietnam war, Captain Benjamin Willard is sent on a dangerous mission that, officially, \"does not exist, nor will it ever exist.\" His goal is to locate - and eliminate - a mysterious Green Beret Colonel named Walter Kurtz, who has been leading his personal army on illegal guerrilla missions into enemy territory.",
            "genres": [
                "Drama",
                "War"
            ],
            "poster": "https://image.tmdb.org/t/p/w500/gQB8Y5RCMkv2zwzFHbUJX3kAhvA.jpg",
            "release_date": 303523200
        },
```

**Expected behavior**
Same order

**MeiliSearch Version**: 0.21.0

**Comment by bidoubiwa** [comment]
To reproduce, you should add this as setting: 
```
{
  "searchableAttributes": [
    "title",
    "description"
  ],
  "filterableAttributes": [
    "genres"
  ]
}
```

**Comment by gmourier** [comment]
Hello!

@MarinPostma (Please, tell me if it's not related to the transplant code), can we keep the order of the fields of the initial document in the response to be ISO with 0.20? 

`searchableAttributes` should not impact this, it would be nice to keep this initial logic in my opinion!

**Comment by MarinPostma** [comment]
Not possible, this is tied to the representation on the documents in milli @Kerollmops

**Comment by Kerollmops** [comment]
Indeed, it can't be returned in the original order, it can only be returned in the `searchableAttribute` order, it is tied to the internal representation of the documents.

**Comment by gmourier** [comment]
Thanks! 

If I understand correctly the user must rewrite all these initially ordered fields in searchableAttributes if he wants to keep the original order.

In this case, he will not search on specific fields and the searchableAttributes feature does not fulfill its initial role.

Should we explain this somewhere to guide users and make it clear that the order can no longer be guaranteed from the moment searchableAttributes operates? A tough question. @dichotommy

**Comment by MarinPostma** [comment]
TBH it should not matter. JSON are manipulated programatically, and the order is irrelevant. We make assumptions on the order of the fields at indexing, but the [JSON spec](https://www.json.org/json-en.html) specifies that a json object is unordered. Implementation of json deserializer may not preserve order.

**Comment by gmourier** [comment]
Thanks for the clarification, @MarinPostma 👍. Indeed, this is something that should be clarified in this way. I think of the user who might wonder why and not being able to find an answer.

**Comment by qdequele** [comment]
@MarinPostma however we use a linked hashmap to keep the order when indexing. As long as the `searchableAttribute` or `displayedAttributes` parameters are not changed the order will remain the same as the first update. @meilisearch/core-team  do you know on what is the return order based on? Why does changing `searchableAttributes` impact the final result?

**Comment by MarinPostma** [comment]
@qdequele the fieldId doubles as field position, that's why

**Comment by bidoubiwa** [comment]
I think this behavior may be a bit problematic. Given the following scenario: 

1. Add settings `filterableAttributes: "genres"`
2. Add documents with the following fields: `id, title, genres, ..`
3. Thus the implicit value of filterableAttributes becomes `* == ["genres", "id", "title",...]`

Now given a concrete example, here is a dataset: 
```js
[
  {
    id: 11,
    title: 'Star Wars',
    overview:
      'Princess Leia is captured and held hostage by the evil Imperial forces in their effort to take over the galactic Empire. Venturesome Luke Skywalker and dashing captain Han Solo team together with the loveable robot duo R2-D2 and C-3PO to rescue the beautiful princess and restore peace and justice in the Empire.',
    genres: ['Adventure', 'Action', 'Science Fiction'],
    poster: 'https://image.tmdb.org/t/p/w500/6FfCtAuVAW8XJjZ7eWeLibRLWTw.jpg',
    release_date: 233366400,
  },
  {
    id: 30,
    title: 'Magnetic Rose',
    overview: '',
    genres: ['Animation', 'Science Fiction'],
    poster: 'https://image.tmdb.org/t/p/w500/gSuHDeWemA1menrwfMRChnSmMVN.jpg',
    release_date: 819676800,
  },
]
```

When doing a search on `query = S` I will receive `Magnetic Rose` instead of `Star Wars`.

**Comment by curquiza** [comment]
Even if the order is relevant in JSON, this behavior can impact the front-end. Cf the mini-dashboard: we clearly see the switch of the fields.
Impact the front-end is not a behavior the users can expect when they update the `searchableAttributes`.

**Comment by bidoubiwa** [comment]
@curquiza Is this not fixed with this meilisearch/milli/pull/293 ?

**Comment by curquiza** [comment]
Nope, it's not in the v0.21.0 milestones. You can check it directly on the issue

I was not there when the decision of removing this from the v0.21.0 was taken, but, by reading this issue, it does not seem to be an easy fix since we don't know how to do it.

**Comment by mou** [comment]
I created a draft PR not only to declare my intention to fix this issue but also to provide a reproduction case to anyone who decided to make their own attempt.

**Comment by mou** [comment]
Oh, This is a tricky one. I spent quite some time locating the cause and still could not figure out the best approach to fix it. So let me explain the logic that leads to undesirable output first:

- order of field names in `searchableAttributes` determines the order of fields `FieldsIdsMap`
- when we set `searchableAttributes`, it triggers reindex
- new representation of the document is created (I do not know why, but I believe it's used in relevancy calculation somehow), and a new document with new fields order is saved in place of the old one
- at this point, we lost the original order of fields
- later, the order of fields in resulted JSON representation is determined by the order of fields in the stored document

The whole week I thought about this issue, and I think, in the current implementation, this could not be strongly identified as a bug. Because there are other cases when the order of fields in the original document is not preserved. For example, if we add the first document with fields `["first", "second"]` and later add another document with fields `["second", "first']`. After that, the order of fields in the second document will always be `["first", "second"]` (until it changed via `searchableAttributes` or `displayableAttributes`). This is because the order of fields is determined by order of the first appearance in the index.

I think this issue can not be solved without the introduction of additional metadata, which will preserve field order disregarding all other settings. But only if you decide this is a bug, not a feature.

**Comment by mou** [comment]
@curquiza please, take a look at the previous comment. I believe this issue requires an additional round of triage.

**Comment by mou** [comment]
Reordering of fields occurs in `milli/src/update/index_documents/transform.rs:545` `remap_index_documents`, which is called from `milli/src/update/settings.rs:269` `reindex`.

**Comment by mou** [comment]
Here are the reproduction steps for the scenario I mentioned:
```
curl -X POST -H "Content-Type: application/json" --data-binary '{"uid": "field_order_test", "primaryKey": "id"}' http://localhost:7700/indexes
curl -X POST -H "Content-Type: application/json" --data-binary '[{"id": 1, "first": "foo", "second": "bar"}]' http://localhost:7700/indexes/field_order_test/documents
curl -X POST -H "Content-Type: application/json" --data-binary '[{"id": 2, "second": "foo", "first": "bar"}]' http://localhost:7700/indexes/field_order_test/documents
curl -s -X GET http://localhost:7700/indexes/field_order_test/documents | jq
{
  "results": [
    {
      "id": 1,
      "first": "foo",
      "second": "bar"
    },
    {
      "id": 2,
      "first": "bar",
      "second": "foo"
    }
  ],
  "offset": 0,
  "limit": 20,
  "total": 2
}
```

**Comment by mou** [comment]
@Kerollmops, so, if you have already decided to reimplement document storage with new requirements, then I do not see any reason for trying to fix this issue in the current architecture. Or I did not understand you well? Also, feel free to discard my draft PR with the test case.

**Comment by Kerollmops** [comment]
Indeed, we want to do a bunch of optimization on the way we store our documents internally to consume less space, and we will evaluate the amount of work required to keep the original order of the fields at this moment.

Thank you very much for your PR, but will we not keep it as you correctly pointed out that we will refactor many of the code parts related to this behavior, and your PR will likely be outdated by then ❤️

**Comment by mou** [comment]
Ok. I will close it then.

**Comment by curquiza** [comment]
Discussed with @Kerollmops: could be done once the Meilisearch DB storage will be re-worked.
